torch/csrc/distributed/c10d/reducer.cpp

```diff
@@ -238,6 +255,16 @@ void Reducer::mark_variable_ready_sparse(VariableIndex index) {
 void Reducer::autograd_hook(VariableIndex index) {
   std::lock_guard<std::mutex> lock(this->mutex_);
 
+  // We would like to call mark_unused_params for every iteration
+  // However it cannot be done here, because we cannot get what forward
+  // returned and cannot reset called_flag.
+  // That's why I want to expose mark_unused_params to front-end.
+  //if (have not called mark_unused_params yet) {
+  //  mark_unused_params(??);
+  //  // Where to reset this flag to false? post-backward hook?
+  //  called_flag = true;
+  //}
```

Comment:
Chatted with @pietern offline yesterday, below is my understanding of the concern and the recommended approach:


#### Concern

The main concern is that we might have a regression on the `no_sync` mode if we don't implement it in this PR. Assume the application runs two iterations with `no_sync` mode:

1. the first iteration touches all parameters
2. the second iteration touches the first 10 parameters in the model

In this case, after exiting the `no_sync` mode, `prepare_for_backward` is called, and DDP hooks would start to trigger allreduce operations. However, only 10 of the autograd hook would fire in the second backward, and DDP would think all remaining parameters are unused and thus set them to None. This would result in incorrect result.

#### Proposal

To address this issue, one approach is to always update the bitmap in all backward passes, regardless of it's `no_sync` or not. It might be something like:

1. In the autograd hook, set `local_used_maps_ [index] = true`;
2. In `finalize_backward` kick off the allreduce on `local_used_maps_` and compute globally unused params
3. In `finalize_backward` after consuming `local_used_maps_`, reset them all to false/0.

One perf optimization might be kicking off the allreduce in `prepare_for_backward`, but that would mean we have to consider unused parameters in two different places:

* autograd hook in `no_sync` mode
* `prepare_for_backward` by walking through the autograd graph.

@pietern please correct me if I am wrong.

Reply:
Make sense to me. Will rework the last commit. Thanks!

Just a minor correction: with the current implementation we don't set the globally unused params to None. We just keep them untouched.

Reply:
Done. Thanks for the suggestion!